onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1174,92 +1634,389 @@ class PlannerImpl {
     }
   }
 
-  // Whether a given NodeArg has fence or not.
-  // If the buffer is reused, need to check whether original OrtValue has fence or not.
-  bool HasFence(const onnxruntime::NodeArg* arg) {
-    bool has_fence = false;
-    if (arg && arg->Exists()) {
-      OrtValueIndex index = Index(arg->Name());
-      AllocPlanPerValue& value_plan = AllocPlan(index);
-
-      has_fence = value_plan.create_fence_if_async;
-      if (value_plan.alloc_kind == AllocKind::kReuse) {
-        // Buffer reused, check original buffer to see if fence is shared.
-        has_fence = has_fence || AllocPlan(value_plan.reused_buffer).create_fence_if_async;
+  // Convert information in execution plan and memory reuse plan into release plan
+  Status GenerateDeallocationPlan() {
+    // 1. build the consumer list for each value
+    std::vector<InlinedVector<NodeIndex>> value_consumers;
+    int num_ml_values = ort_value_name_idx_map_.MaxIdx() + 1;
+    value_consumers.resize(num_ml_values);
+
+    // iterate each stream from back, so the first element is the last consumer in single stream case
+    for (auto& stream : stream_nodes_) {
+      for (auto it = stream.rbegin(), end = stream.rend(); it != end; ++it) {
+        NodeIndex node_index = *it;
+        auto* node = graph_viewer_.GetNode(node_index);
+
+        auto process_input = [&](const NodeArg& input, size_t /*arg_idx*/) {
+          if (input.Exists()) {
+            const auto& name = input.Name();
+            int value_idx;
+            ORT_RETURN_IF_ERROR(ort_value_name_idx_map_.GetIdx(name, value_idx));
+            auto origin = Buffer(value_idx);
+            if (origin != -1 && plan_.allocation_plan[origin].alloc_kind == AllocKind::kAllocate) {
+              // add current node as consumer for origin buffer
+              value_consumers[origin].push_back(node_index);
+            }
+          }
+          return Status::OK();
+        };
+
+        ORT_RETURN_IF_ERROR(Node::ForEachWithIndex(node->InputDefs(), process_input));
+        ORT_RETURN_IF_ERROR(Node::ForEachWithIndex(node->ImplicitInputDefs(), process_input));
+      }
+    }
+    // 2. build the release actions and fill into node's release list
+    auto process_consumer = [&](size_t release_action_idx, NodeIndex node_index) {
+      plan_.release_actions[release_action_idx].ref_count++;
+      plan_.node_release_list[node_index].push_back(release_action_idx);
+    };
+    plan_.node_release_list.resize(graph_viewer_.MaxNodeIndex() + 1);
+    for (size_t i = 0; i < value_consumers.size(); ++i) {
+      if (!value_consumers[i].empty()) {
+        plan_.release_actions.push_back(SequentialExecutionPlan::ReleaseAction{i, 0});
+        auto release_action_idx = plan_.release_actions.size() - 1;
```

Comment:
I think maybe map is a better data structure for release_action than vector, then we can get rid of release_action_index and the code will be easier to understand. Are you concerned with the perf?